scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java

```diff
@@ -665,14 +662,18 @@ private Windows narrowByStateConstraints(
       return ret;
     }
 
-    //REVIEW: could be some optimization in constraint ordering
+    final var totalDomain = Window.between(windows.minTimePoint().get(), windows.maxTimePoint().get());
+    //make sure the simulation results cover the domain
+    simulationFacade.computeSimulationResultsUntil(totalDomain.end);
 
     //iteratively narrow the windows from each constraint
+    //REVIEW: could be some optimization in constraint ordering (smallest domain first to fail fast)
     for (final var constraint : constraints) {
-      ret = constraint.findWindows(plan, ret);
-      assert ret != null;
+      //REVIEW: loop through windows more efficient than enveloppe(windows) ?
+      final var validity = constraint.evaluate(simulationFacade.getLatestConstraintSimulationResults(), totalDomain, simulationFacade.getLatestEnvironment());
```

Comment:
🔻Since the `environment` is a mapping of activity aliases to activity instances (where the activity aliases come from `ForEachActivity`), the only valid environment to pass at the top level is `Map.of()`.

```suggestion
      final var validity = constraint.evaluate(simulationFacade.getLatestConstraintSimulationResults(), totalDomain, Map.of());
```

Perhaps this is worth adding a new default `evaluate` method to the `Expression` interface, which takes only the first two arguments, and supplies `Map.of()` itself.

Reply:
I have added the default method.